File: code/STEP2/flags.f
      parameter (lshort=7,slplim=1.,slpx=.5)
      character*105 title,cc*3
      sumch=0.
      sumchp=0.
      nsta=0
      nstap=0
      nyrs=0
      nyrsp=0
      nshort=0
      nsl1=0
      nsl2=0
      ndsl=0
      nok=0
      nokx=0
      nswitch=0
      nsw0=0
      open(78,file='work/fort.78',form='formatted')
      open( 2,file='work/fort.2',form='formatted')
   10 read(78,'(a)',end=100) title
      read(title,'(a3,i9,2f9.3,i5,5f9.3,i5,1x,i4,i5,1x,i4)',end=100)
     *    cc,id,sl1,sl2,knee,yk,sl,ylin,rms,rms0,iy1,iy2,iy1e,iy2e
      nsta=nsta+1
      sumch=sumch+sl*(iy2-iy1+1)
      nyrs=nyrs+(iy2-iy1+1)
      if(sl.lt.0.) then
        nstap=nstap+1
        nyrsp=nyrsp+(iy2-iy1+1)
        sumchp=sumchp+sl*(iy2-iy1+1)
      end if
C***  classify : iflag: +1 for short legs etc
      iflag=0
      if(knee.lt.iy1+lshort.or.knee.gt.iy2-lshort) iflag=iflag+1
      if(knee.lt.iy1+lshort.or.knee.gt.iy2-lshort) nshort=nshort+1
      if(abs(sl1).gt.slplim) iflag=iflag+20
      if(abs(sl1).gt.slplim) nsl1=nsl1+1
      if(abs(sl2).gt.slplim) iflag=iflag+10
      if(abs(sl2).gt.slplim) nsl2=nsl2+1
      if(abs(sl2-sl1).gt.slplim) iflag=iflag+100
      if(abs(sl2-sl1).gt.slplim) ndsl=ndsl+1
      if(abs(sl2-sl1).gt.slpx) iflag=iflag+100
      if(iflag.eq.0) nok=nok+1
      if(iflag.eq.100) nokx=nokx+1
      if(sl1*sl2.lt.0..and.abs(sl1).gt..2.and.
     *  abs(sl2).gt..2) iflag=iflag+1000
      if(iflag.ge.1000) nswitch=nswitch+1
      if(iflag.eq.1000) nsw0=nsw0+1
      write(title(101:105),'(i5)') iflag
      write(2,'(a)') title
      go to 10

  100 write (*,*) 'all',nsta,-sumch/nsta,-10.*sumch/nyrs
      write (*,*) 'urb warm',nstap,-sumchp/nstap,-10.*sumchp/nyrsp
      write(*,*) '# short,sl1,sl2,dsl,ok',nshort,nsl1,nsl2,ndsl,nok,nokx
      write (*,*) 'switches: all , else ok ',nswitch,nsw0
      stop
      end

File: code/flags.py
#!/usr/bin/env python
"""Python replacement for code/STEP2/flags.f

Input files:

    work/fort.78

Output files:
    work/fort.2

The input file is a text file, with a single record per line. This is created
using the ``PApars.py`` script. This script adds the flags value at the end of 
each record, based on the record's contents. (Later the output file is given
a header line to create ``work/PApars.list``.)

"""
__docformat__ = "restructuredtext"


import ccc_binary
import script_support


# Constants
lshort = 7
slplim = 1.0
slpx = 0.5


def readRec(f):
    """Read a single record from the input file.

    The input file is text, line oriented, one record per line.

    :Param f:
        The open file to read from.
    """
    l = f.readline()
    l = l.rstrip()
    if not l:
        return None, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0

    (CCdStationID,slope_l,slope_r,knee,Yknee,slope,Ymid,RMS,RMSl,
            rur_urb,ext_range) = l.split()
    cc, IDc = CCdStationID[:3], int(CCdStationID[3:])
    sl1 = float(slope_l)
    sl2 = float(slope_r)
    knee = int(knee)
    sl0 = float(slope)
    iy1, iy2 = [int(v) for v in rur_urb.split("-")]
    iy1e, iy2e = [int(v) for v in ext_range.split("-")]
    return cc, IDc, sl1, sl2, knee, Yknee, sl0, Ymid,RMS,RMSl, iy1, iy2, iy1e, iy2e, l


def main(args):
    """The main for this module.

    """
    f78 = open("work/PApars.pre-flags")
    f2 = open("work/PApars.list", "w")
    f2.write('CCdStationID  slope-l  slope-r knee    Yknee    slope     Ymid      RMS     RMSl 3-rur+urb ext.range flag\n')
    nstat = sumch = nyrs = nstap = nyrsp = sumchp = 0
    nsl1 = nsl2 = ndsl = nswitch = nsw0 = nok = nokx = nsta = 0
    nshort = 0
    while True:
        data = readRec(f78)
        if data[0] is None:
            break
        cc,id,sl1,sl2,knee,yk,sl,ylin,rms,rms0,iy1,iy2,iy1e,iy2e,line = data
        nsta += 1
        sumch = sumch + sl * (iy2 - iy1 + 1)
        nyrs = nyrs + (iy2 - iy1 + 1)
        if sl < 0.0:
            nstap += 1
            nyrsp += (iy2 - iy1 + 1)
            sumchp += sl * (iy2 - iy1 + 1)

        # classify : iflag: +1 for short legs etc
        iflag = 0
        if knee < iy1+lshort or knee > iy2 - lshort:
            iflag += 1
        if knee < iy1+lshort or knee > iy2 - lshort:
            nshort += 1
        if abs(sl1) > slplim:
            iflag += 20
        if abs(sl1) > slplim:
            nsl1 += 1
        if abs(sl2) > slplim:
            iflag += 10
        if abs(sl2) > slplim:
            nsl2 += 1
        if abs(sl2 - sl1) > slplim:
            iflag += 100
        if abs(sl2 - sl1) > slplim:
            ndsl += 1
        # TODO: The small offset fixes a Fortran/Python rounding issue, but I
        #       am not happy with it. PAO, Mon Feb 23 2009.
        if abs(sl2 - sl1) + 0.00000001 > slpx:
            iflag += 100
       
        if iflag == 0:
            nok += 1
        if iflag == 100:
            nokx += 1
        if sl1 * sl2 < 0.0 and abs(sl1) > 0.2 and abs(sl2) > 0.2:
            iflag += 1000
        if iflag >= 1000:
            nswitch += 1
        if iflag == 1000:
            nsw0 += 1
       
        line += " %4d" % iflag
        f2.write("%s\n" % line)

    print " %-10s %4d %10.7f     %10.7f" % ("all", nsta,-sumch/nsta,-10.*sumch/nyrs)
    print " %-11s %8d  %10.7f    %10.7f" % ("urb warm", nstap,-sumchp/nstap,-10.*sumchp/nyrsp)
    print " %-11s %11d %11d %11d %11d %11d %11d" % (
        "# short,sl1,sl2,dsl,ok", nshort,nsl1,nsl2,ndsl,nok,nokx)
    print " %-11s  %11d %11d" % ("switches: all , else ok", nswitch,nsw0)


if __name__ == "__main__":
    import optparse
    usage = "usage: %prog [options]"
    parser = script_support.makeParser(usage)
    options, args = script_support.parseArgs(parser, __doc__, (0, 0))
    main(args)

